core/store/src/migrations.rs

```diff
@@ -29,23 +43,57 @@ impl<'a> BatchedStoreUpdate<'a> {
         Ok(())
     }
 
-    pub fn set_ser<T: BorshSerialize>(
+    fn set_or_insert_ser<T: BorshSerialize>(
         &mut self,
         col: DBCol,
         key: &[u8],
         value: &T,
+        insert: bool,
     ) -> std::io::Result<()> {
         let value_bytes = value.try_to_vec()?;
-        self.batch_size += key.as_ref().len() + value_bytes.len() + 8;
-        self.store_update.as_mut().unwrap().set(col, key.as_ref(), &value_bytes);
+        let entry_size = key.as_ref().len() + value_bytes.len() + 8;
+        self.batch_size += entry_size;
+        self.total_size_written += entry_size as u64;
+        let update = self.store_update.as_mut().unwrap();
+        if insert {
+            update.insert(col, key.as_ref(), &value_bytes);
+        } else {
+            update.set(col, key.as_ref(), &value_bytes);
+        }
 
         if self.batch_size > self.batch_size_limit {
             self.commit()?;
         }
+        if self.total_size_written - self.printed_total_size_written > PRINT_PROGRESS_EVERY_BYTES {
+            info!(
+                target: "migrations",
+                "Migrations: {} written",
+                bytesize::to_string(self.total_size_written, true)
+            );
+            self.printed_total_size_written = self.total_size_written;
+        }
```

Comment:
Should we do this only on commit?

Reply:
Meh, the commit is once per 10MB and print is once per 1GB so it doesn't really make much of a difference. It's just to show the user it isn't stuck, anyway.